src/irr/asset/CGLILoader.cpp

```diff
@@ -0,0 +1,157 @@
+/*
+MIT License
+Copyright (c) 2019 AnastaZIuk
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+#include "CBufferLoaderBIN.h"
+
+//#ifdef _IRR_COMPILE_WITH_GLI_
+#include "gli/gli.hpp"
+//#else
+//	#error "Need GLI"
+//#endif
+
+namespace irr
+{
+	namespace asset
+	{
+		asset::SAssetBundle CBufferLoaderBIN::loadAsset(io::IReadFile* _file, const asset::IAssetLoader::SAssetLoadParams& _params, asset::IAssetLoader::IAssetLoaderOverride* _override, uint32_t _hierarchyLevel)
+		{
+			if (!_file)
+				return {};
+
+			SContext ctx(_file->getSize());
+			ctx.file = _file;
+
+			const char* filename = _file->getFileName().c_str();
+			gli::texture texture = gli::load(filename);
+			if (texture.empty())
+			{
+				os::Printer::log("Failed to load texture at path ", filename, ELL_ERROR);
+				return {};
+			}
+			
+		    gli::gl GL(gli::gl::PROFILE_GL33);
+			gli::gl::format const format = GL.translate(texture.format(), texture.swizzles());
+			GLenum target = GL.translate(texture.target());
+			
+			/*GLuint textureName = 0;
+			glGenTextures(1, &textureName);
+			glBindTexture(target, textureName);
+			glTexParameteri(target, GL_TEXTURE_BASE_LEVEL, 0);
+			glTexParameteri(target, GL_TEXTURE_MAX_LEVEL, static_cast<GLint>(Texture.levels() - 1));
+			glTexParameteri(target, GL_TEXTURE_SWIZZLE_R, format.Swizzles[0]);
+			glTexParameteri(target, GL_TEXTURE_SWIZZLE_G, format.Swizzles[1]);
+			glTexParameteri(target, GL_TEXTURE_SWIZZLE_B, format.Swizzles[2]);
+			glTexParameteri(target, GL_TEXTURE_SWIZZLE_A, format.Swizzles[3]);*/
+			
+			glm::tvec3<GLsizei> const extent(texture.extent()); // it is on 0 level without specyfing it explicitly
+			GLsizei const totalFaces = static_cast<GLsizei>(texture.layers() * texture.faces());
+
+			ICPUImage::SCreationParams imageInfo;
+			imageInfo.format = format;
```

Comment:
you will need to translate the format between gli format and our own E_FORMAT

Reply:
@devshgraphicsprogramming well, so the **format** struct looks like this: 
https://github.com/g-truc/gli/blob/master/gli/gl.hpp#L335

I will translate it, but I'm not sure which format member I should choose for translating
internal format: https://github.com/g-truc/gli/blob/master/gli/gl.hpp#L16
external format: https://github.com/g-truc/gli/blob/master/gli/gl.hpp#L238
or type format: https://github.com/g-truc/gli/blob/master/gli/gl.hpp#L265

Reply:
@devshgraphicsprogramming what do you think about it?

Reply:
Thanks to `GL_ARB_texture_storage` we now deal with exclusively internal formats
https://github.com/g-truc/gli/blob/master/gli/gl.hpp#L16

Before OpenGL was a retarded mix of extenal+type format, i.e. `GL_RGBA` + `GL_FLOAT`

Now you just have `GL_RGBA32F`

You should ignore swizzles and target types (well you still need to differentiate 1D, 2D and 3D) for Images and put them in the ImageView creation struct instead, arbitrary swizzles are possible.